Implemented a class Quaternion

NikhilPappu opened the pull request:
The class allows one to instantiate quaternions as objects and provides
basic quaternion operations.
I placed the class in polys/domains/quaternion.py as there isn't any
module specifically named 'algebra'.

<!-- Please give this pull request a descriptive title. Pull requests with descriptive titles are more likely to receive reviews. Describe what you changed! A title that only references an issue number is not descriptive. -->

<!-- If this pull request fixes an issue please indicate which issue by typing "Fixes #NNNN" below. -->

jksuom commented:
This does not seem to fit very well in `polys.domains` since that is dedicated to so-called *domains*. In SymPy, those are commutative rings that inherit a number attributes from `Domain`. They are intended to function as coefficient rings for various algebras.

There is currently no module for non-commutative algebra. Apparently, most (or all) algebras have been commutative so far. It is probably too early to create a new algebra module only for quaternions. Instead, I think this could be placed in `polys.agca` for now (even though `ca` stands for 'commutative algebra').

NikhilPappu commented:
@jksuom Thanks for the suggestion. I shall place it in polys.agca. I think i need to add a test file first. I got some errors in Travis.

Upabjojr commented on sympy/polys/domains/quaternion.py:
```diff
@@ -0,0 +1,524 @@
+
+
+class Quaternion(Basic):
+
```
If you are going to insert it into a formula, it should inherit Expr

Upabjojr commented on sympy/polys/domains/quaternion.py:
```diff
@@ -0,0 +1,524 @@
+        q = self
+        # trigsimp is used to simplify sin(x)^2 + cos(x)^2 (these terms
+        # arise when from_axis_angle is used).
+        return sqrt(trigsimp(q.a ** 2 + q.b ** 2 + q.c ** 2 + q.d ** 2))
```
Beware that trigsimp may slowdown the code by a lot.

Upabjojr commented on sympy/polys/domains/quaternion.py:
```diff
@@ -0,0 +1,524 @@
+                          q1.b * q2.c - q1.c * q2.b + q1.d * q2.a
+                          + q1.a * q2.d)
+
+    def conj(self):
```
I think there is a SymPy default method for conjugation, will check later.

Upabjojr commented on sympy/polys/domains/quaternion.py:
```diff
@@ -0,0 +1,524 @@
+        q = self
+        return q * (1 / q.norm())
+
+    def inverse(self):
```
I think we use inv() for matrices, shall we keep that same naming?

Upabjojr commented on sympy/polys/domains/quaternion.py:
```diff
@@ -0,0 +1,524 @@
+        return Matrix([[m00, m01, m02, m03], [m10, m11, m12, m13],
+                       [m20, m21, m22, m23], [m30, m31, m32, m33]])
+
+class CustomLatexPrinter(LatexPrinter):
```
Why a custom printer?

Upabjojr commented on sympy/polys/domains/quaternion.py:
```diff
@@ -0,0 +1,524 @@
+        return str(expr.a) + '\n+ ' + '(' + str(expr.b) + ')i\n+ ' \
+            + '(' + str(expr.c) + ')j\n+ ' + '(' + str(expr.d) + ')k'
+
+Basic.__str__ = lambda self: CustomStrPrinter().doprint(self)
```
Still is a bad idea

NikhilPappu replied:
Thanks for all the suggestions. I shall inherit from Expr and rename inverse to inv. I really need cos(x)^2 + sin(x)^2 to simply to 1 though. Do you know of a way I can implement a form of trigsimp that checks only for this?

As for the printing, I knew all along that it was a bad idea but I wasn't sure what to do.
I guess it is clear to me now.

I shall remove the custom printers and add the _print_Quaternion methods to StrPrinter , LatexPrinter etc.

NikhilPappu replied:
Can u tell me where exactly I should add tests. Should I add them in sympy.core.tests/test_args or in sympy.polys.agca/tests or in both?

Upabjojr replied:
This class should not be located in `sympy/polys`, maybe in `sympy/physics`? or in `sympy/geometry`?

Upabjojr commented on sympy/polys/domains/quaternion.py:
```diff
@@ -0,0 +1,524 @@
+        return latex(expr.a) + '+' + latex(expr.b * i) + '+' + \
+            latex(expr.c * j) + '+' + latex(expr.d * k)
+
+class CustomStrPrinter(StrPrinter):
```
Why not just add this to sympy.printing.StrPrinter?

Upabjojr commented on sympy/polys/domains/quaternion.py:
```diff
@@ -0,0 +1,524 @@
+        q2 = other
+
+        # If q1 or q2 is a number or a sympy expression instead of a quaternion
+        if not isinstance(q1, Quaternion):
```
you are assuming that `q1` is a complex, it might not be the case.

NikhilPappu replied:
I did this because I thought it would be better to take a general case. If real is set to true for an expression then im() will be 0.

Upabjojr replied:
What if `q1` is a matrix or a Grassmann number?

NikhilPappu replied:
So would it be a good idea to consider the default case to be real and have a special case for complex?

jksuom replied:
It looks like `q1 = self` would be an object of the `Quaternion` class.

NikhilPappu replied:
@jksuom 
Thanks for pointing that out. I should've seen that. I guess I should just check for q2 then.
One problem that I'm facing is that q*4 is working but 4*q isn't. Do u know how to fix this?

jksuom replied:
There is a priority mechanism implemented in [core](https://github.com/sympy/sympy/blob/master/sympy/core/expr.py#L95) for that purpose. If a quaternion q has higher priority than a complex number c, then a `c*q` will be computed as `q.__rmul__(c)` instead of `c.__mul__(q)` (and it is possible to define `__rmul__ = __mul__`).

NikhilPappu replied:
@jksuom 
Thanks a lot! I used __radd__ and __rmul__ in my code and it worked.

Upabjojr replied:
Anyway, I would add:
```python
if not q2.is_complex:
    return Add(self, q2)
```
or something similar, that is leave `quaterion + noncomplex` unevaluated.

Upabjojr commented on sympy/polys/domains/quaternion.py:
```diff
@@ -0,0 +1,524 @@
+        d=0,
+    ):
+
+        self.a = a
```
These should be transformed into a SymPy object.

NikhilPappu replied:
Would it be enough to use simpify()?

Upabjojr replied:
the function `a = sympify(a)`

Upabjojr commented:
This looks nice, the location is wrong.

Some work is still needed to make it better interact with SymPy. We should discuss about the location of this code.

NikhilPappu committed:
Implemented a class Quaternion

The class allows one to instantiate quaternions as objects and provides
basic quaternion operations.
I placed the class in polys/domains/quaternion.py as there isn't any
module specifically named 'algebra'.

Upabjojr committed:
refactoring class Quaternion

Upabjojr commented:
@NikhilPappu I made some edits, because it would have taken too long to explain everything.

I put the code in `sympy/core`, I'm not sure about this decision, as the _core_ is supposed to be for the core functionality of SymPy only.

jksuom commented:
I don't think this should be in core. At some point we should have a module for algebras, and I think that would be the proper place for this. Currently, `polys` is where all algebraic stuff is located although the name seems to refer to only a part of that. Perhaps we could define `sympy.algebras` now. At this point, `sympy.polys.agca` is the only module containing algebra type structures (module structure in addition to ring structure) although they cannot be directly applied to quaternions.

jksuom commented on sympy/core/quaternion.py:
```diff
@@ -169,12 +185,13 @@ def add(self, other):
         if not isinstance(q2, Quaternion):
-            return Quaternion(re(q2) + q1.a, im(q2) + q1.b, q1.c, q1.d)
+            if q2.is_complex:
+                return Quaternion(re(q2) + q1.a, im(q2) + q1.b, q1.c, q1.d)
```
Here we could probably have `q1._a`, etc. to save an extra call.

jksuom commented on sympy/printing/str.py:
```diff
@@ -713,6 +713,13 @@ def _print_Complement(self, expr):
+    def _print_Quaternion(self, expr):
+        i, j, k = symbols('i j k')
+        # TODO: This expression is potentially confusing,
+        # shall we print it as `Quaternion( ... )`?
```
I think that is what `ReprPrinter` should do.

NikhilPappu replied:
I think u mixed up the latex and str code.

Upabjojr replied:
Oh, yes. Sorry about that. I just wanted to show where to put the code.

NikhilPappu commented on sympy/core/quaternion.py:
```diff
@@ -340,11 +355,12 @@ def pow_cos_sin(self, p):
-    def diff(self, *args):
-        return Quaternion(diff(self.a, *args), diff(self.b, *args),
-                          diff(self.c, *args), diff(self.d, *args))
+    def _eval_derivative(self, x):
```
I had already added _eval_Integral an _eval_diff on line 161. I let those methods call diff and integrate.

Upabjojr replied:
Sorry I didn't notice that

Upabjojr commented:
@NikhilPappu do you have some usage test scripts? I believe you should put them in the test file, so that Travis can verify the correct behaviour of `Quaternion`.

Upabjojr committed:
moved to sympy.algebra

Upabjojr commented:
I have moved the script to the newly created module _sympy.algebra_.

I made some edits to the code and to the printer.

A test file has been started. We need many more tests.

jksuom commented on setup.py:
```diff
@@ -70,6 +70,7 @@
 # Check that this list is uptodate against the result of the command:
 # python bin/generate_module_list.py
 modules = [
+    'sympy.algebra',
```
I think that the plural form `algebras` would be more suitable.

Upabjojr replied:
I have been thinking for a while, both could be OK. I prefer to avoid plurals to make things simpler.

Upabjojr replied:
But maybe in this case it's more appropriate to use the plural, as there will be probably many algebras.

jksuom replied:
That is what I had in mind, too. It should also tell that not *all* algebraic matters are going to be included.

jksuom replied:
We could also have `Complex` built on the same model: two component expressions with the familiar multiplication rules. That should not be confused with `ComplexFloat` that could exist in `core.numbers` and have `mpc` as representation.

NikhilPappu commented:
@Upabjojr I have a few test cases. Ill write more of them and add them to the test file. 
I'm sorry if I am not being very active. I have exams coming up.
Can you tell me why the Travis build failed and what needs to be done for it to pass?

jksuom commented:
New modules and tests should be added to the lists in `setup.py`. See `bin/test_setup.py`.

Upabjojr committed:
renamed algebra to algebras

jksuom commented:
Would it be possible to write `_copysign(x, y)` in the following form?

    if y == 0:
        return 0  # or maybe S(0)
    return x if x*y > 0 else -x

NikhilPappu commented:
I didn't think of that. That would be a much better implementation of copysign.

Upabjojr commented:
How should this class behave in case we construct `Quaterion(3+4*I, 0, 0, 0)`? Shall it return `Quaternion(3, 4, 0, 0)`?

NikhilPappu commented:
I don't think that would be a good idea. I think we should just check the parameters and disallow complex expressions

Upabjojr commented on sympy/algebras/tests/test_quaternion.py:
```diff
@@ -0,0 +1,90 @@
+
+
+    
+    
```
only one new line at the end of the file, otherwise tests will fail.

jksuom replied:
More precisely, only one newline character.

NikhilPappu replied:
Sorry about that. I forgot to use autopep8.

Upabjojr commented on sympy/algebras/quaternion.py:
```diff
@@ -39,13 +39,19 @@ def __new__(cls, a=0, b=0, c=0, d=0):
         d = sympify(d)
 
-        obj = Expr.__new__(cls, a, b, c, d)
+        if((not(a.is_real) and not(a.is_real is None))
```
A bit too restrictive, I would allow the case `a.is_real == None`.

NikhilPappu replied:
This check allows one to create a Quaternion(x, x, x, x)
if 
x = symbols('x') or if x = symbols('x', real = True)
but not if
x = symbols('x', real = False) (as in Quaternion(3 + 4*I, 0, 0, 0))

I feel it would be cumbersome for the user to set real = True everytime.

Upabjojr commented on sympy/algebras/quaternion.py:
```diff
@@ -39,13 +39,19 @@ def __new__(cls, a=0, b=0, c=0, d=0):
+        if((not(a.is_real) and not(a.is_real is None))
+                or (not(b.is_real) and not(b.is_real is None))
+                or (not(c.is_real) and not(c.is_real is None))
+                or (not(d.is_real) and not(d.is_real is None))):
```
it would be clearer with `if any(i.is_real == False for i in [a, b, c, d]):`

Upabjojr commented on sympy/algebras/quaternion.py:
```diff
@@ -39,13 +39,19 @@ def __new__(cls, a=0, b=0, c=0, d=0):
+                or (not(b.is_real) and not(b.is_real is None))
+                or (not(c.is_real) and not(c.is_real is None))
+                or (not(d.is_real) and not(d.is_real is None))):
+            print("Invalid arguments passed.")
```
`raise ValueError("arguments have to be real")`

Upabjojr commented on sympy/algebras/quaternion.py:
```diff
@@ -124,10 +130,10 @@ def __copysign(x, y):
         if y == 0:
             return 0
-        return x if x*y > 0 else -x
+        return x if x * y > 0 else -x
```
PEP8 tools keep putting spaces between `*` symbols, but sometimes it's clearer to read expressions such as `x*y`, especially when you have mixed _+_ and _-_ symbols. It depends on the case.

Upabjojr commented on sympy/algebras/quaternion.py:
```diff
@@ -0,0 +1,534 @@
+        c = sympify(c)
+        d = sympify(d)
+
+        if(any(not(i.is_real) and not(i.is_real is None) for i in [a, b, c, d])):
```
```python
In [1]: True == False
Out[1]: False

In [2]: None == False
Out[2]: False

In [3]: False == False
Out[3]: True
```

I believe `i.is_real == False` would be easier to read.

NikhilPappu replied:
Thanks for pointing that out. I have changed it.

asmeurer commented:
I agree a new submodule makes the most sense. It doesn't really fit anywhere else.

NikhilPappu committed:
Made some minor changes and added some tests

NikhilPappu committed:
Minor change

NikhilPappu committed:
Added checks to disallow complex arguments

NikhilPappu committed:
Removed some unnecessary spaces

NikhilPappu committed:
Minor change

jksuom commented on sympy/algebras/quaternion.py:
```diff
@@ -0,0 +1,534 @@
+        c = sympify(c)
+        d = sympify(d)
+
+        if any(i.is_real is False for i in [a, b, c, d]):
```
I'm not sure that checking for reality is necessary. It should be enough that the coordinates are commutative. It makes perfect sense to define a quaternion algebra over any commutative ring, only it need not be a (skew) field in general.

NikhilPappu replied:
@jksuom I see your point. I'll change 
```is_real``` to ```is_commutative```.

jksuom commented on sympy/algebras/quaternion.py:
```diff
@@ -39,7 +39,7 @@ def __new__(cls, a=0, b=0, c=0, d=0):
 
-        if any(i.is_real is False for i in [a, b, c, d]):
+        if any(i.is_commutative is False for i in [a, b, c, d]):
             raise ValueError("arguments have to be real")
```
real -> commutative

NikhilPappu replied:
I didn't get you. U suggested just checking for commutative right?

jksuom replied:
In the error message as well.

NikhilPappu replied:
Sorry about that. I changed it.

Upabjojr commented:
> It makes perfect sense to define a quaternion algebra over any commutative ring, only it need not be a (skew) field in general.

There's a small problem with this: the quaternion operations such as _mul_ and _add_ are defined as to interact with SymPy reals and complexes. So if you have a quaternion over the complex numbers and you add a complex number to it, you get the addition of the real and imaginary parts in the 1st and 2nd fields of the Quaternion. I suppose that a quaternion over the complex field should add it to the first position only.

It would be nice if Python supported something like C++'s templates, so you can define the field upon which the quaternion is acting, but we don't have such capabilities.

The computer algebra system (Axiom)[https://en.wikipedia.org/wiki/Axiom_(computer_algebra_system)] tries to define the algebraic groups and algebras by using a template paradigm in their own (very strange) programming language.

jksuom commented:
> a quaternion over the complex field should add it to the first position only.

This is what I think too.  The field of complex numbers can be embedded into (standard) quaternions in many ways (infinite number, in fact). It may be useful in some cases to consider the (natural looking) embedding `I -> i` but it probably should not be automatic.

NikhilPappu commented:
I made some changes for addition of complex numbers with complex elements. If you think the code is fine then we can add something similar for multiplication.

asmeurer commented on sympy/algebras/quaternion.py:
```diff
@@ -0,0 +1,545 @@
+
+    is_commutative = False
+
+    def __new__(cls, a=0, b=0, c=0, d=0, real_elements = True):
```
Please document the real_elements flag. 

From what I understand, it differentiates whether `i`, the quaternion, is treated as sqrt(-1). I think a better name could be used. How about `i_is_complex=False`.

NikhilPappu replied:
I was just asking if this is a good way to achieve add and mul in quaternions over the real numbers and quaternions over the complex numbers.

Upabjojr replied:
what about `field=S.Reals` instead?

NikhilPappu replied:
@Upabjojr I think that would be a better name. Should I let the values if the elements influence this flag or should it be left to the user to change it?

asmeurer commented on sympy/algebras/quaternion.py:
```diff
@@ -0,0 +1,545 @@
+        return self._c
+
+    @property
+    def d(self):
```
Are a, b, c, d standard when working with quaternions? It seems it would get confusing to keep track of which corresponds to the usual i, j, k.

NikhilPappu replied:
The Wikipedia page uses a, b, c, d in the definition.
https://en.wikipedia.org/wiki/Quaternion

asmeurer replied:
But are these incidental or commonly used? You have to use something as the scalar variables, but that doesn't necessarily mean the variable names are part of the definition. 

I don't use quaternions myself, so I don't know. Do people immediately associate in their heads something like c <=> j?

NikhilPappu replied:
The variable names are not part of the definition.
The two most commonly used conventions I came across are
1. a + b*i + c*j + d*k
2. w + x*i + y*j + z*k

asmeurer commented:
How about adding i, j, k constants to the module namespace. Except I'm not sure what to call them. Perhaps i_, j_, k_. I would suggest I, J, K, but then I conflicts with the complex I. Plain lowercase i, j, k are no good because they are used too often in Python as loop variables.

NikhilPappu commented:
@asmeurer 
I don't think it would be necessary to create constants i_, j_, k_ because quaternions are not as commonly used or as fundamental as complex numbers. I think dealing with quaternions using a class avoids any confusion.

Upabjojr commented:
> I think dealing with quaternions using a class avoids any confusion.

I also think the same.

asmeurer commented:
Well presently the str printer uses a form that doesn't recreate the expression, unless you manually define `i`, `j`, and `k`. So it would be useful at least for that. If we don't add basis elements, I think the str printer should print "Quaternion(a, b, c, d)" (the pretty printer can still print the nice form, since that isn't used to recreate expressions). 

I also don't see why you think that people wouldn't want to create quaternions using the basis elements, like `1 + 2*i_ + 3*j_ + 4*k_` (instead of `Quaternion(1, 2, 3, 4)`).

asmeurer commented:
We could also add i, j, k as class variables to Quaternion (they wouldn't even need to be prefixed in that case, since they would be namespaced).

Upabjojr commented:
> Well presently the str printer uses a form that doesn't recreate the expression, unless you manually define i, j, and k.

I know, I would make it print `Quaternion( ... )` but there was some disagreement.

The argument against the expanded representation is brought on by quaternions over fields: how do you specify the field they are acting upon?

asmeurer commented:
I'm not sure I'm following. Is this still the question of whether the quaternion i should equal sqrt(-1)? I think this is a different question from the base field.

NikhilPappu commented:
@asmeurer 
>I'm not sure that checking for reality is necessary. It should be enough that the coordinates are commutative. It makes perfect sense to define a quaternion algebra over any commutative ring, only it need not be a (skew) field in general.

> There's a small problem with this: the quaternion operations such as mul and add are defined as to interact with SymPy reals and complexes. So if you have a quaternion over the complex numbers and you add a complex number to it, you get the addition of the real and imaginary parts in the 1st and 2nd fields of the Quaternion. I suppose that a quaternion over the complex field should add it to the first position only.
It would be nice if Python supported something like C++'s templates, so you can define the field upon which the quaternion is acting, but we don't have such capabilities.
The computer algebra system Axiom tries to define the algebraic groups and algebras by using a template paradigm in their own (very strange) programming language.

Upabjojr commented:
Anyway, this is how quaternions are implemented in FriCAS/Axiom:
https://github.com/fricas/fricas/blob/3ad41812a0e0ef202a8a29391ff3d7d121501177/src/algebra/quat.spad#L15

this is how the addition is implemented in Boost's C++ Quaternion:
https://github.com/boostorg/math/blob/242e9d669198415d165bb0213f730c00985bee9a/include/boost/math/quaternion.hpp#L207

Apparently in Boost the arguments of the quaternion are assumed to be real, despite employing a templated.

FriCAS/Axiom support generic commutative rings, though I haven't found the code to handle the interaction with scalars.

jksuom commented:
> the code to handle the interaction with scalars.

I think that comes from Module via [Algebra](https://github.com/fricas/fricas/blob/master/src/algebra/catdef.spad#L122).

Upabjojr commented:
> I think that comes from Module via Algebra.

How do you think they handle addition of a complex number and a quaternion over the real field?

I don't have it installed and cannot verify.

NikhilPappu committed:
Changed is_real to is_commutative

NikhilPappu committed:
Minor change

NikhilPappu committed:
Changed addition for quaternions with complex elements

NikhilPappu committed:
Changed add and mul for quaternions over complex fields

NikhilPappu committed:
Small fix

jksuom commented:
I believe that to add a complex number and a quaternion in Axiom the number has to be explicitly converted to a quaternion. The compiler can probably do that automatically only when a suitable coercion has been defined. In quad.spad there seems to exist only two such coercions, one for elements of the coefficient ring of the quaternions and another for integers that are first converted to elements of the coefficient ring.

       coerce(r : R)      ==
         quatern(r, 0$R, 0$R, 0$R)
       coerce(n : Integer)      ==
         quatern(n :: R, 0$R, 0$R, 0$R)

Complex algebras are defined in gaussian.spad, but I could not find anything in the code that would connect those to quaternions.

NikhilPappu commented:
@asmeurer How can I go about adding i_, j_, k_?

Upabjojr commented:
@NikhilPappu I would suggest to keep things simple. That edit may make this PR even more complicated.

Upabjojr commented on sympy/algebras/tests/test_quaternion.py:
```diff
@@ -0,0 +1,98 @@
+    ((2 * sqrt(29)/29, 3 * sqrt(29)/29, 4 * sqrt(29)/29),
+                                   2 * acos(sqrt(30)/30))
+
+    assert q1.to_matrix() == Matrix([[-2/5, 2/5, 11/5],
```
we should use SymPy integers, so `-S(2)/5` instead of `-2/5`

NikhilPappu commented:
> @NikhilPappu I would suggest to keep things simple. That edit may make this PR even more complicated.

I guess you are right.

Upabjojr commented:
The method `.to_matrix()` is ambiguous, it should be clear that you want to represent a rotation matrix, maybe it should be called `.to_rotation_matrix()`.

Anyway, how does it work?

I was expecting to get a rotation around the _z_-axis, but instead I got this:
```
In [25]: q = Quaternion(cos(theta/2), 0, 0, sin(theta/2))

In [26]: q.to_matrix()
Out[26]: 
⎡      _________       _________                                    ⎤
⎢     ╱    2⎛θ⎞       ╱    2⎛θ⎞                ⎛θ⎞    ⎛θ⎞           ⎥
⎢-   ╱  sin ⎜─⎟  +   ╱  cos ⎜─⎟          -2⋅sin⎜─⎟⋅cos⎜─⎟           ⎥
⎢  ╲╱       ⎝2⎠    ╲╱       ⎝2⎠                ⎝2⎠    ⎝2⎠           ⎥
⎢───────────────────────────────   ─────────────────────────────   0⎥
⎢     _________       _________        _________       _________    ⎥
⎢    ╱    2⎛θ⎞       ╱    2⎛θ⎞        ╱    2⎛θ⎞       ╱    2⎛θ⎞     ⎥
⎢   ╱  sin ⎜─⎟  +   ╱  cos ⎜─⎟       ╱  sin ⎜─⎟  +   ╱  cos ⎜─⎟     ⎥
⎢ ╲╱       ⎝2⎠    ╲╱       ⎝2⎠     ╲╱       ⎝2⎠    ╲╱       ⎝2⎠     ⎥
⎢                                                                   ⎥
⎢                                       _________       _________   ⎥
⎢             ⎛θ⎞    ⎛θ⎞               ╱    2⎛θ⎞       ╱    2⎛θ⎞    ⎥
⎢        2⋅sin⎜─⎟⋅cos⎜─⎟          -   ╱  sin ⎜─⎟  +   ╱  cos ⎜─⎟    ⎥
⎢             ⎝2⎠    ⎝2⎠            ╲╱       ⎝2⎠    ╲╱       ⎝2⎠    ⎥
⎢ ─────────────────────────────   ───────────────────────────────  0⎥
⎢     _________       _________        _________       _________    ⎥
⎢    ╱    2⎛θ⎞       ╱    2⎛θ⎞        ╱    2⎛θ⎞       ╱    2⎛θ⎞     ⎥
⎢   ╱  sin ⎜─⎟  +   ╱  cos ⎜─⎟       ╱  sin ⎜─⎟  +   ╱  cos ⎜─⎟     ⎥
⎢ ╲╱       ⎝2⎠    ╲╱       ⎝2⎠     ╲╱       ⎝2⎠    ╲╱       ⎝2⎠     ⎥
⎢                                                                   ⎥
⎣               0                                0                 1⎦
```

Upabjojr commented:
I mean, the matrix is not even orthogonal:

```
In [38]: ao.T*ao
Out[38]: 
⎡                                2                                                                      ⎤
⎢     1                     2⋅sin (θ)                                                                   ⎥
⎢─────────── + ───────────────────────────────────                          0                          0⎥
⎢   2⎛θ   π⎞                                     2                                                      ⎥
⎢tan ⎜─ + ─⎟   ⎛  _____________     ____________⎞                                                       ⎥
⎢    ⎝2   4⎠   ⎝╲╱ -cos(θ) + 1  + ╲╱ cos(θ) + 1 ⎠                                                       ⎥
⎢                                                                                                       ⎥
⎢                                                                                   2                   ⎥
⎢                                                        1                     2⋅sin (θ)                ⎥
⎢                        0                          ─────────── + ───────────────────────────────────  0⎥
⎢                                                      2⎛θ   π⎞                                     2   ⎥
⎢                                                   tan ⎜─ + ─⎟   ⎛  _____________     ____________⎞    ⎥
⎢                                                       ⎝2   4⎠   ⎝╲╱ -cos(θ) + 1  + ╲╱ cos(θ) + 1 ⎠    ⎥
⎢                                                                                                       ⎥
⎣                        0                                                  0                          1⎦
```

NikhilPappu commented:
@Upabjojr I'm sorry about that issue. I figured out the right way to do it. I shall implement it the right way tomorrow.

Upabjojr commented:
Just copy from Wikipedia:
https://en.wikipedia.org/wiki/Quaternions_and_spatial_rotation

NikhilPappu committed:
Added docstrings for complex field examples

Removed the code that checks for real elements and changes the
real_field tag. I thought it would be best if only the user had control
over the real_field tag.

NikhilPappu committed:
Small fix

NikhilPappu committed:
Fixed to_rotation_matrix and changed doctests and tests

NikhilPappu committed:
Small change

NikhilPappu committed:
Small change

Upabjojr commented:
```python
In [17]: var("theta", real=True)
Out[17]: θ

In [18]: q = Quaternion(cos(theta/2), 0, 0, sin(theta/2))

In [19]: q.to_axis_angle()
Out[19]: 
⎛⎛          ⎛θ⎞ ⎞                ⎞
⎜⎜       sin⎜─⎟ ⎟                ⎟
⎜⎜          ⎝2⎠ ⎟        ⎛   ⎛θ⎞⎞⎟
⎜⎜0, 0, ────────⎟, 2⋅acos⎜cos⎜─⎟⎟⎟
⎜⎜      │   ⎛θ⎞│⎟        ⎝   ⎝2⎠⎠⎟
⎜⎜      │sin⎜─⎟│⎟                ⎟
⎝⎝      │   ⎝2⎠│⎠                ⎠
```

I was wondering, can we get the exact value `((0, 0, 1), theta)`?

Upabjojr committed:
added rotational test

Upabjojr commented:
> I don't think it is easy to get the exact value.

The expression looks as if the result is uncertain about the axis:

```
⎛          ⎛θ⎞ ⎞
⎜       sin⎜─⎟ ⎟
⎜          ⎝2⎠ ⎟
⎜0, 0, ────────⎟
⎜      │   ⎛θ⎞│⎟
⎜      │sin⎜─⎟│⎟
⎝      │   ⎝2⎠│⎠
```
If `theta` is positive the _z_-axis is positive, otherwise it's negative. This does not really make sense. I would expect a definite axis. Maybe there's some better algorithm the generate the axis-angle pair?

Upabjojr commented:
I think that this issue could be actually left for the future.

Upabjojr commented on sympy/algebras/quaternion.py:
```diff
@@ -0,0 +1,540 @@
+        return cls(a, b, c, d).normalize()
+
+    @classmethod
+    def from_matrix(cls, M):
```
this should be called `from_rotation_matrix`, i.e. there are many possible quaternion-matrix relationships.

Upabjojr commented on sympy/algebras/quaternion.py:
```diff
@@ -0,0 +1,540 @@
+
+        return Quaternion(a, b, c, d)
+
+    def ln(self):
```
This needs a better name, ideally we would like to call `log(q)`, but that probably requires some hacks. Maybe let's keep it private for the time (prepend a `_` sign)?

Upabjojr commented on sympy/algebras/quaternion.py:
```diff
@@ -0,0 +1,540 @@
+                          integrate(self.c, *args), integrate(self.d, *args))
+
+    @staticmethod
+    def rotate(pin, r):
```
Maybe this should be called _rotate triplet_ or something more specific? `rotate` is a bit vague as a name.

Upabjojr commented on sympy/algebras/tests/test_quaternion.py:
```diff
@@ -0,0 +1,105 @@
+    assert conjugate(q) == Quaternion(x, -y, -z, -w)
+    assert q.norm() == sqrt(w**2 + x**2 + y**2 + z**2)
+    assert q.normalize() == Quaternion(x, y, z, w) / sqrt(w**2 + x**2 + y**2 + z**2)
+    assert q.inverse() == Quaternion(x, -y, -z, -w) / (w**2 + x**2 + y**2 + z**2)
```
are we missing the test for the power of a quaternion?

Upabjojr commented:
OK, I've added some last comments, after that I think this PR is almost ready for merging.

Upabjojr commented on sympy/algebras/tests/test_quaternion.py:
```diff
@@ -53,19 +53,14 @@ def test_quaternion_functions():
-    Quaternion(log(sqrt(30)),
-    2 * sqrt(29) * acos(sqrt(30)/30) / 29,
-    3 * sqrt(29) * acos(sqrt(30)/30) / 29,
-    4 * sqrt(29) * acos(sqrt(30)/30) / 29)
```
Let's keep these tests.

NikhilPappu committed:
Minor changes

NikhilPappu committed:
Merge branch 'quaternion' of github.com:NikhilPappu/sympy into quaternion

NikhilPappu committed:
Small change

Upabjojr commented:
OK, now it's good to merge.

Thanks for your contribution!

NikhilPappu commented:
I was able to make my first open source contribution because of you guys. Thanks a lot for helping me out!

Upabjojr commented:
> I was able to make my first open source contribution because of you guys. Thanks a lot for helping me out!

It was a really nice work for being your first contribution to SymPy.

NikhilPappu commented:
> It was a really nice work for being your first contribution to SymPy.
Thank You!

RohitJain3241 commented:
I want to implement some more properties and applications to quaternions. Any suggestions would be really appreciated. @Upabjojr @asmeurer
